src/combinatorics/graph/multi/basic.lean

```diff
@@ -0,0 +1,136 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Antoine Labelle. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Antoine Labelle
+-/
+import category_theory.limits.shapes.terminal
+
+/-!
+# Indexed multigraphs
+
+This file defines (indexed) multigraphs. A multigraph is a collection of vertices and a collection
+of edges with two maps from the edges to the vertices (representing their ends) and a compatible
+involution of the edges.
+
+We provide the category structure and show that multigraphs have a terminal object, namely the graph
+with one vertex and one half-loop.
+
+## Main declarations
+
+* `multigraph`: Multigraphs. Also the category of multigraphs.
+* `multigraph.discrete`: For a type `X`, the graph with a single half-loop at `x : X` and no other
+  edge.
+
+## References
+
+* [Evan Patterson, *Graphs and C-sets*](https://www.algebraicjulia.org/blog/post/2020/09/cset-graphs-1/)
+-/
+
+open category_theory category_theory.limits
+
+universes u v
+
+/-- A multigraph is a type of vertices `V`, a type of edges `E`, two maps `E → V` representing the
+edges ends, and an involution of the edges that respects the ends.
+
+This definition allows *half-loops*, edges from a vertex `v` to itself that are fixed under the
+edges involution. -/
+@[nolint check_univs]
+structure multigraph :=
+(V : Type u)
+(E : Type v)
+(fst snd : E → V)
+(inv : E → E)
+(inv_fst : ∀ e, fst (inv e) = snd e . obviously)
+(inv_snd : ∀ e, snd (inv e) = fst e . obviously)
+(inv_inv : ∀ e, inv (inv e) = e . obviously)
```

Comment:
I'm a bit out of the loop and trying to catch up so bear with me - I'm finding that exposing `V` makes it really tricky to talk about subgraphs because of the different coercions going on. I'm running into that a bit when I try to move between `spanning_coe` and `coe` to say anything about cycles - there's API for `spanning_coe` but none for `subgraph.coe`.

Reply:
I can't see how bundling the type would make your life any easier in thsi situation. You would still have the coercions.